File: backend/routers/publicaciones.py
"""
publicaciones.py — Las publicaciones del catálogo para la pestaña Omnicanal.

  GET /api/publicaciones            → página filtrable + censo de cobertura
  GET /api/publicaciones/estados    → qué estados existen en cada canal
  GET /api/publicaciones/cobertura  → solo el censo, sin traer las filas

LECTURA PURA, con UNA excepción declarada: `GET /api/publicaciones?refrescar=true`
—y sólo con ese parámetro, que por default viene apagado— le pregunta a Mercado
Libre el precio de las publicaciones del SKU exacto de `q` y lo guarda antes de
leer (`services/precio_al_abrir.py`). Todo lo demás de este router no escribe en
kubera, no habla con ningún marketplace y no toca ningún flujo vivo.

Toda la interpretación (qué es "activa" en cada canal, cuándo hay oferta, cuándo
se puede calcular el margen) vive en `services/publicaciones_panel.py`, con el
porqué de cada regla y el censo que la respalda.

Las seis cosas que hay que saber antes de pintar esto:

  1. El margen es PROSPECTIVO (contra el precio que la publicación cobra hoy),
     no realizado. El panel de Análisis contesta la otra pregunta.
  2. `margen_pct: null` significa "no se puede saber", con `margen_motivo` al
     lado. NUNCA se devuelve 0 por falta de datos.
  3. `oferta_estado` tiene TRES valores. `desconocida` no es `sin_oferta`.
  4. Una oferta puede estar SIN CONFIRMAR (`oferta_confirmada: false`): se
     observó antes del último cambio de la publicación, así que nadie sabe si
     sigue viva. Esas NO se aplican — `precio_vigente` y el margen van contra
     el precio de lista, y `oferta_precio` / `oferta_desc_pct` llegan en null.
     Lo observado no se pierde: viaja en `oferta_precio_visto` /
     `oferta_desc_pct_visto` para pintarlo marcado. El censo trae el conteo en
     `cobertura.ofertas_sin_confirmar` (665 de 665 el 25-ago-2026).
  5. Hay TRES precios y ninguno se llama como uno esperaría (v0.262.0):

       precio_lista   lo que ML TACHA          `listings.price_base`
       precio_ml      el precio del VENDEDOR   `listings.price`
                      tras SUS campañas
       precio_vigente  lo que el comprador PAGA `listings.price_sale` si está
                       confirmado; si no, cae a `precio_ml` y viene MARCADO

     `precio_vigente_confirmado` (bool | null) dice si ese último número se
     puede creer. `null` = no aplica (canal sin capa de promoción).
     `cobertura.precio_sin_confirmar` lleva el conteo (789 de 806 activas de ML
     el 25-ago, antes de abrir `public_offers`; baja solo con los avisos).
  6. **Un margen puede venir MARCADO** — y ésta es la combinación nueva de
     v0.262.0, la que puede tumbar a quien asuma el contrato viejo:

       margen_pct     el número. NO cambia.
       margen_motivo  por qué NO hay margen. Sigue apareciendo SOLO cuando
                      `margen_pct` es null. **Esa regla no cambió.**
       margen_aviso   NUEVO. "sí hay número, pero tómalo con reserva".
                      Vocabulario cerrado; hoy un solo valor:
                      "precio_sin_confirmar". Viene CON `margen_pct` presente.
       margen_contra  NUEVO. contra qué precio se calculó:
                      "precio_cobrado" (price_sale confirmado) o
                      "precio_ml" (el techo conocido, sin confirmar).

     Medido: `price` sobreestima lo que ML cobra en 1.44x de mediana (p90 2.95,
     máx 4.85; 60 publicaciones vivas, 25-ago-2026). Por eso se marca. El trato
     es el MISMO que ya recibe el costo implausible en `frontend/lib/margen.ts`:
     ámbar y ⚠, no el rojo/verde que se lee como un hecho. El censo trae el
     conteo por canal en `cobertura.canales[].avisos` — aparte de `motivos`,
     porque una fila marcada SÍ cuenta en `con_margen`.
"""
from __future__ import annotations

import logging

from fastapi import APIRouter, HTTPException, Query

from services import precio_al_abrir
from services import publicaciones_panel as pp

log = logging.getLogger("omnicanal.routers.publicaciones")
router = APIRouter(prefix="/api/publicaciones", tags=["publicaciones"])


@router.get("")
async def listar(
    canal: str | None = Query(None, description="mercado_libre | amazon | tiktok | temu | walmart | general"),
    tienda: str | None = Query(None, description="legacy_code de la cuenta: BEKURA, SANCORFASHION, AMAZON…"),
    estado: str | None = Query(None, description="activa | puede_estar_activa | no_comprable | pausada | en_revision | borrador | rechazada | cerrada | sin_estado | desconocido"),
    solo_activas: bool = Query(False, description="atajo: estado in (activa, puede_estar_activa)"),
    solo_con_oferta: bool = Query(False),
    q: str | None = Query(None, description="busca en sku, título y listing_id"),
    orden: str = Query("sku", description="sku | precio_desc | precio_asc | margen_desc | margen_asc | descuento_desc | reciente"),
    page: int = Query(1, ge=1),
    per_page: int = Query(50, ge=1, le=500),
    refrescar: bool = Query(False, description="confirma contra ML el precio de las publicaciones del SKU EXACTO que venga en `q`, antes de contestar"),
):
    """
    Una página de publicaciones + el censo de cobertura de ESE mismo filtro.

    El censo viaja junto a los datos a propósito: un margen promedio sin la
    banda de cobertura al lado se lee como un hecho sobre todo el catálogo, y
    hoy solo lo es sobre poco más de la mitad de Mercado Libre.

    `refrescar=true` — LO ÚNICO QUE ESCRIBE EN ESTE ROUTER, y sólo aquí. Antes
    de leer, le pregunta a Mercado Libre el precio que cobra por las
    publicaciones del SKU **exactamente igual** a `q` y lo guarda en
    `channel.listings.price_sale`. Es lo que hace que el cajón del producto
    muestre lo que la tienda cobra en ese momento y no una foto de hace horas.

    Tres cosas que hay que saber antes de mandarlo:

      1. **Va con el SKU exacto, no con una búsqueda.** El objetivo se arma con
         `l.sku = q` (citext, o sea sin distinguir mayúsculas). Una `q` que sea
         parte de un título o de un `listing_id` refresca CERO publicaciones,
         no todas las que la búsqueda encuentre. Es a propósito: recorrer una
         lista no puede convertirse en cientos de llamadas a ML.
      2. **Nunca rompe la pantalla.** Si ML falla, tarda o no hay token, la
         lectura sigue igual con lo guardado y el bloque `refresco` dice qué
         pasó. Un cajón que no abre es peor que uno con un dato de hace una
         hora.
      3. **Lo que hay que mirar es `refresco.al_dia`** (bool): "todo lo que se
         muestra de ML para este SKU está confirmado contra ML ahora mismo".
         Cuando es `false`, los precios de ML del cajón son los guardados y hay
         que decirlo en pantalla — `precio_vigente_confirmado` y `oferta_dias`
         siguen contando la verdad publicación por publicación. `estado` trae
         el porqué con vocabulario cerrado: `ok` (se preguntó), `piso` (se
         observó hace menos de PRECIO_AL_ABRIR_PISO_MIN, no hacía falta),
         `sin_publicaciones`, `apagado`, `no_aplica`, `sin_token`, `fallo`,
         `timeout`.
    """
    try:
        refresco = await precio_al_abrir.refrescar_sku(q) if refrescar else None
        res = await _a_hilo(
            pp.listar, canal=canal, cuenta=tienda, estado=estado,
            solo_activas=solo_activas, solo_con_oferta=solo_con_oferta,
            search=q, orden=orden, page=page, per_page=per_page)
        if refresco is not None:
            res["refresco"] = refresco
        return res
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    except Exception as exc:  # noqa: BLE001
        log.exception("publicaciones.listar falló")
        raise HTTPException(status_code=502,
                            detail=f"No se pudo leer publicaciones: {exc}") from exc


@router.get("/estados")
async def estados(canal: str | None = Query(None)):
    """
    El censo de estados por canal, con el valor CRUDO de cada uno al lado.

    Existe para que la pestaña nunca pinte un 0 sin explicación: si un canal no
    reporta estado utilizable —Temu contesta códigos, Walmart solo publica en
    `status`— aquí sale `sin_estado`/`puede_estar_activa` con su `nota`, en vez
    de una lista vacía que se leería como "no hay publicaciones activas".
    """
    try:
        return {"canales": await _a_hilo(pp.censo_estados, canal=canal)}
    except Exception as exc:  # noqa: BLE001
        log.exception("publicaciones.estados falló")
        raise HTTPException(status_code=502, detail=str(exc)) from exc


@router.get("/cobertura")
async def cobertura(canal: str | None = Query(None),
                    tienda: str | None = Query(None)):
    """El censo solo, sin las filas. Para el encabezado de la pestaña."""
    try:
        res = await _a_hilo(pp.listar, canal=canal, cuenta=tienda,
                            page=1, per_page=1)
        return res["cobertura"]
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc
    except Exception as exc:  # noqa: BLE001
        log.exception("publicaciones.cobertura falló")
        raise HTTPException(status_code=502, detail=str(exc)) from exc


async def _a_hilo(fn, **kw):
    """
    Regla 11 de la casa: `supabase_db` es psycopg2, o sea BLOQUEANTE. Una
    lectura de ~8,700 filas dentro del event loop congelaría los webhooks de ML
    mientras dura. Sale a un hilo.
    """
    import asyncio
    return await asyncio.to_thread(lambda: fn(**kw))

File: frontend/lib/publicaciones.ts
/**
 * publicaciones.ts — El vocabulario de la pestaña Omnicanal, y SOLO eso.
 *
 * Aquí se PRESENTA lo que el backend ya decidió. No se recalcula un margen, no
 * se deduce una oferta y no se infiere si una publicación está activa: las tres
 * cosas se miden canal por canal en `services/publicaciones_panel.py`, contra el
 * censo real, y llegan resueltas. Este archivo elige palabras y colores.
 *
 * Las tres trampas que este módulo existe para no repetir:
 *
 *  1. "Activa" NO es un valor único. ML dice `active`, TikTok `ACTIVATE`,
 *     Walmart `PUBLISHED`, Temu contesta códigos numéricos y Amazon distingue
 *     `BUYABLE` de `DISCOVERABLE`. Se pinta el estado NORMALIZADO; el crudo solo
 *     aparece cuando el backend no supo mapearlo (`desconocido`).
 *  2. Una oferta NO es "el campo de precio de campaña viene lleno". De 3,029
 *     filas con `price_sale` poblado, solo 609 traen un precio MENOR: pintar por
 *     campo-no-vacío mostraría 2,420 descuentos que no existen.
 *  3. `margen_pct: null` es "no se puede saber", no 0%. Amazon, TikTok, Temu y
 *     Walmart no tienen costo propio y el backend NO manda margen para ellos a
 *     propósito, para que nadie lo derive con la comisión equivocada.
 */

import type {
  EstadoNormalizado,
  MargenMotivo,
  OfertaEstado,
  RefrescoEstado,
  RefrescoPrecio,
} from "./types";

/** Tono visual de un estado. El color concreto lo pone el componente. */
export type TonoEstado = "vivo" | "vivo_matizado" | "ambar" | "gris";

export interface EstadoUI {
  label: string;
  tono: TonoEstado;
  /** Qué significa exactamente. Va en el tooltip; no se resume más. */
  ayuda: string;
}

/**
 * El vocabulario CERRADO del backend. Si aparece una llave nueva aquí sin que
 * el backend la mande, sobra; si el backend manda una que no está, cae en
 * `desconocido` y se muestra el crudo (que es justo lo que hay que ver).
 */
export const ESTADO_UI: Record<EstadoNormalizado, EstadoUI> = {
  activa: {
    label: "Activa",
    tono: "vivo",
    ayuda: "Se puede comprar AHORA.",
  },
  puede_estar_activa: {
    label: "Puede estar activa",
    tono: "vivo_matizado",
    ayuda:
      "El canal no distingue activo de inactivo, así que no se puede afirmar. "
      + "Es lo más cerca de la verdad que ese canal permite.",
  },
  no_comprable: {
    label: "No comprable",
    tono: "ambar",
    ayuda:
      "Existe y se ve en el catálogo, pero NO se vende. No es activa ni pausada: "
      + "es su propia cosa (Amazon DISCOVERABLE).",
  },
  pausada: { label: "Pausada", tono: "gris", ayuda: "Existe y está apagada." },
  en_revision: {
    label: "En revisión",
    tono: "gris",
    ayuda: "El canal la está revisando; todavía no vende.",
  },
  borrador: { label: "Borrador", tono: "gris", ayuda: "Nunca se publicó." },
  rechazada: {
    label: "Rechazada",
    tono: "gris",
    ayuda: "El canal la rechazó.",
  },
  cerrada: { label: "Cerrada", tono: "gris", ayuda: "Se dio de baja." },
  sin_estado: {
    label: "Sin estado",
    tono: "gris",
    ayuda:
      "El canal NO reporta estado para esta publicación. No significa que no "
      + "haya publicaciones: significa que no se sabe.",
  },
  desconocido: {
    label: "Desconocido",
    tono: "gris",
    ayuda:
      "Valor nuevo del canal que todavía nadie mapeó. Se muestra tal cual llegó "
      + "para que se pueda mapear, en vez de aplastarlo a otra cosa.",
  },
};

/** `estado` normalizado → ¿cuenta como viva para el filtro? (igual que el backend). */
export const ESTADOS_VIVOS: EstadoNormalizado[] = ["activa", "puede_estar_activa"];

/** Orden en que se ofrecen los estados en el filtro. */
export const ESTADOS_ORDEN: EstadoNormalizado[] = [
  "activa",
  "puede_estar_activa",
  "no_comprable",
  "pausada",
  "en_revision",
  "borrador",
  "rechazada",
  "cerrada",
  "sin_estado",
  "desconocido",
];

// ── Oferta ───────────────────────────────────────────────────────────────────

export const OFERTA_UI: Record<OfertaEstado, { label: string; ayuda: string }> = {
  con_oferta: {
    label: "Con oferta",
    ayuda: "El precio de campaña es MENOR que el de lista: hay descuento vivo.",
  },
  sin_oferta: {
    label: "Sin oferta",
    ayuda: "Se observó el precio de campaña y NO había promoción.",
  },
  desconocida: {
    label: "Sin observar",
    ayuda:
      "Nadie le ha preguntado al canal por el precio de campaña de esta "
      + "publicación. NO es lo mismo que 'sin oferta'.",
  },
};

/**
 * A partir de cuántos días una oferta deja de poder leerse como "de hoy".
 *
 * Quien escribe la observación (`precios_venta.py`) está dormido: la más nueva
 * en producción es del 21-ago. Una oferta sin su antigüedad al lado se lee como
 * vigente y puede llevar días muerta.
 */
export const OFERTA_DIAS_AMBAR = 2;

// ── Margen ───────────────────────────────────────────────────────────────────

/**
 * Por qué NO se puede saber el margen. Nunca se sustituye por 0 ni por un
 * guion mudo: el motivo es la mitad de la información.
 */
export const MOTIVO_TEXTO: Record<MargenMotivo | "desconocido", string> = {
  sin_costo_del_canal:
    "No hay costo de ESTE canal. Hoy solo Mercado Libre tiene motor de costos; "
    + "el costo de ML no se hereda a otro canal a propósito, porque su comisión "
    + "es distinta y el margen saldría falso.",
  sin_comision:
    "Falta el porcentaje de comisión de la categoría: sin él no hay cómo restar "
    + "lo que se lleva el canal.",
  sin_peso:
    "Falta el peso. Sin peso el costeo mete 0.5 kg de oficio y el fee de envío "
    + "sale del renglón barato de la tarifa: el margen quedaría optimista sin avisar.",
  sin_precio: "La publicación no tiene precio.",
  desconocido: "El backend no dijo por qué.",
};

/** Texto corto para el chip del censo. */
export const MOTIVO_CORTO: Record<MargenMotivo | "desconocido", string> = {
  sin_costo_del_canal: "sin costo del canal",
  sin_comision: "sin comisión",
  sin_peso: "sin peso",
  sin_precio: "sin precio",
  desconocido: "sin motivo",
};

// ── Confirmación del precio al abrir el cajón ────────────────────────────────

/**
 * Por qué NO se pudo confirmar el precio contra Mercado Libre. Sólo se usan
 * los estados que dejan `al_dia: false`; `ok`, `piso` y `sin_publicaciones` no
 * necesitan explicación porque no se avisa nada cuando todo está al día.
 */
const REFRESCO_MOTIVO: Partial<Record<RefrescoEstado, string>> = {
  apagado: "la confirmación en vivo está apagada",
  sin_token: "falta el token de esa cuenta de Mercado Libre",
  fallo: "Mercado Libre no contestó",
  timeout: "Mercado Libre tardó más de la cuenta",
  no_aplica: "no se pidió la confirmación",
};

/**
 * El porqué, en palabras de quien lo lee. Llamarlo SOLO cuando
 * `refresco.al_dia === false`.
 *
 * El caso mixto es el que importa y por eso va primero: `estado: "ok"` con
 * `sin_respuesta > 0` significa que a unas publicaciones ML sí les contestó y a
 * otras no. El aviso habla de las que NO, que son las que obligan a leer el
 * número con reserva.
 *
 * Todos los campos menos `al_dia` son opcionales en el contrato, así que aquí
 * no se asume que ninguno venga.
 */
export function motivoRefresco(r: RefrescoPrecio): string {
  if (r.sin_respuesta && r.sin_respuesta > 0) {
    return r.sin_respuesta === 1
      ? "una publicación no contestó"
      : `${r.sin_respuesta} publicaciones no contestaron`;
  }
  if (r.omitidas_tope && r.omitidas_tope > 0) {
    return "se alcanzó el tope de consultas por producto";
  }
  return (
    (r.estado && REFRESCO_MOTIVO[r.estado])
    ?? "Mercado Libre no confirmó el precio"
  );
}

// ── Canales y tiendas ────────────────────────────────────────────────────────

export const CANAL_LABEL: Record<string, string> = {
  mercado_libre: "Mercado Libre",
  amazon: "Amazon",
  tiktok: "TikTok",
  temu: "Temu",
  walmart: "Walmart",
  general: "WooCommerce",
};

export function labelCanal(canal: string): string {
  return CANAL_LABEL[canal] ?? canal;
}

export interface TiendaOpcion {
  /** `legacy_code`, tal cual lo espera el parámetro `tienda` del backend. */
  id: string;
  label: string;
  canal: string;
}

/**
 * Las tiendas por canal.
 *
 * Solo Mercado Libre opera dos cuentas; el resto tiene una y ahí "canal" y
 * "tienda" coinciden. Los `legacy_code` salen del contrato del backend y se
 * verificaron contra los datos: si alguno cambiara, el filtro devolvería 0
 * filas —no un error—, así que la lista se comprueba, no se supone.
 */
export const TIENDAS: TiendaOpcion[] = [
  { id: "BEKURA", label: "Kubera", canal: "mercado_libre" },
  { id: "SANCORFASHION", label: "San Corpe", canal: "mercado_libre" },
  { id: "AMAZON", label: "Amazon (San Corpe)", canal: "amazon" },
  { id: "KUBERA", label: "TikTok Shop", canal: "tiktok" },
  { id: "TEMU", label: "Temu", canal: "temu" },
  { id: "WALMART", label: "Walmart", canal: "walmart" },
];

export function tiendasDe(canal: string | null): TiendaOpcion[] {
  return canal ? TIENDAS.filter((t) => t.canal === canal) : TIENDAS;
}

export function labelTienda(id: string | null | undefined): string {
  if (!id) return "—";
  return TIENDAS.find((t) => t.id === id)?.label ?? id;
}

// ── Formato ──────────────────────────────────────────────────────────────────
//
// Se formatea, no se recalcula. La verdad del número vive en el backend.

const MXN = new Intl.NumberFormat("es-MX", {
  style: "currency",
  currency: "MXN",
  maximumFractionDigits: 2,
});

export function fmtMoneda(v: number | null | undefined, moneda = "MXN"): string {
  if (v === null || v === undefined || !Number.isFinite(v)) return "—";
  if (moneda && moneda !== "MXN") {
    return new Intl.NumberFormat("es-MX", {
      style: "currency",
      currency: moneda,
      maximumFractionDigits: 2,
    }).format(v);
  }
  return MXN.format(v);
}

/** Fracción (0.458) → "45.8 %". `null` NO se convierte en "0 %". */
export function fmtPct(v: number | null | undefined, decimales = 1): string {
  if (v === null || v === undefined || !Number.isFinite(v)) return "—";
  return `${(v * 100).toFixed(decimales)} %`;
}

/** Igual, pero con el signo delante: un margen negativo tiene que leerse negativo. */
export function fmtPctFirmado(v: number | null | undefined, decimales = 1): string {
  if (v === null || v === undefined || !Number.isFinite(v)) return "—";
  const s = (v * 100).toFixed(decimales);
  return `${v > 0 ? "+" : ""}${s} %`;
}

export function fmtEntero(v: number | null | undefined): string {
  if (v === null || v === undefined || !Number.isFinite(v)) return "—";
  return new Intl.NumberFormat("es-MX").format(v);
}

/** "hace 4.1 días" / "hace 3 h". El dato viene en días con decimal. */
export function fmtAntiguedad(dias: number | null | undefined): string {
  if (dias === null || dias === undefined || !Number.isFinite(dias)) return "sin fecha";
  if (dias < 1) {
    const h = Math.round(dias * 24);
    return h <= 1 ? "hace menos de 1 h" : `hace ${h} h`;
  }
  const d = dias < 10 ? dias.toFixed(1) : Math.round(dias).toString();
  return `hace ${d} d`;
}
